File: 2022/day01/part2.py
def check(best: list[int], curr: int) -> list[int]:
    i = -1
    while i != len(best) - 1 and curr > best[i + 1]:
        i += 1
    if i == -1:
        pass
    else:
        best = best[1 : i + 1] + [curr] + best[i + 1 :]
    return best


def solution():
    best = [0, 0, 0]
    curr = 0
    with open("day01/input.txt") as f:
        for line in f:
            clean_line = line.strip()
            print(clean_line)
            if not clean_line:
                best = check(best, curr)
                curr = 0
            else:
                curr += int(clean_line)
        # Check the last line as well
        best = check(best, curr)
    print(f"{best=}, {sum(best)=}")


solution()

File: 2023/day01/src/bin/part2.rs
#[allow(dead_code)]
const INPUT: &'static str = "\
two1nine
eightwothree
abcone2threexyz
xtwone3four
4nineeightseven2
zoneight234
7pqrstsixteen";

#[allow(dead_code)]
const EXPECTED: &'static str = "281";

fn main() {
    let input = include_str!("../input.txt");
    let sol = solution(input);
    dbg!(sol);
}

fn solution(s: &str) -> String {
    let nums = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];

    let ans: u32 = s.lines().map(|line| {
        let mut matches: Vec<u32> = Vec::new();
        for (i, c) in line.chars().enumerate() {
            if let Some(x) = c.to_digit(10) {
                matches.push(x);
            } else {
                for (j, num) in nums.iter().enumerate() {
                    if line[i..].starts_with(num) {
                        matches.push((j + 1) as u32);
                        break;
                    }
                }
            }
        }
        matches.first().unwrap() * 10 + matches.last().unwrap()
    }).sum();
    format!("{}", ans)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_solution() {
        assert!(solution(INPUT) == EXPECTED.to_string());
    }
}
